crates/pallet-domains/src/staking_epoch.rs

```diff
@@ -45,21 +57,128 @@ fn switch_operator<T: Config>(operator_id: OperatorId) -> Result<(), Error> {
     })
 }
 
+pub(crate) fn do_unlock_operators<T: Config>(
+    domain_id: DomainId,
+    domain_block_number: T::DomainNumber,
+) {
+    if let Some(operators) = PendingOperatorUnlocks::<T>::take(domain_id, domain_block_number) {
+        for operator_id in operators {
+            if let Err(err) = unlock_operator::<T>(operator_id) {
+                error!("Failed to unlock operator pool[{operator_id:?}]: {err:?}",)
+            }
+        }
+    }
+}
+
+fn unlock_operator<T: Config>(operator_id: OperatorId) -> Result<(), Error> {
+    Operators::<T>::try_mutate(operator_id, |maybe_operator| {
+        let operator = maybe_operator.as_mut().ok_or(Error::MissingOperator)?;
+
+        let mut total_shares = operator.total_shares;
+        let mut total_stake = operator
```

Comment:
the `total_stake` here will act as `remaining_stake` at the end. As for the `total_shares`, not sure what you mean here. We would still need track the remaining shares to calculate the nominator's stake. Am I missing something here ?

Reply:
What I mean was something like this:
```rust
        let total_shares = operator.total_shares;
        let total_stake = operator
            .current_total_stake
            .checked_add(&operator.current_epoch_rewards)
            .ok_or(TransitionError::BalanceOverflow)?;
        let mut remaining_stake = total_stake;

        let freeze_identifier = T::FreezeIdentifier::staking_freeze_id(operator_id);

        Nominators::<T>::drain_prefix(operator_id).try_for_each(|(nominator_id, nominator)| {
            let nominator_share = Perbill::from_rational(nominator.shares, total_shares);
            let nominator_staked_amount = nominator_share.mul_floor(total_stake);

            let locked_amount = T::Currency::balance_frozen(&freeze_identifier, &nominator_id);
            let amount_to_mint = nominator_staked_amount
                .checked_sub(&locked_amount)
                .unwrap_or(Zero::zero());

            // remove the lock and mint any gains
            ...

            // update pool's remaining stake
            remaining_stake = remaining_stake
                .checked_sub(&nominator_staked_amount)
                .ok_or(TransitionError::BalanceUnderflow)?;

            // remove any pending deposits for this nominator and then remove the lock
            ...

            Ok(())
        })?;
```

Currently, we updating `total_stake` and `total_share` and using them to calculate `nominator_staked_amount` at the same time, it is not straightforward to see the calculation is correct IMO.

Reply:
That is incorrect. You suggestion does not account for shares that are burned. If you don't update the shares, then `nominator_share` will be always be incorrect.

Reply:
What do you mean by burned shares? If there are 10 total shares and 2 nominators with 3 and 7 shares respectively, the `nominator_share` of them should be `3/10` and `7/10`, right?

Reply:
I understand what you are saying now. It was done intentionally to always calculate share portion with latest `total_stake` and `total_share` considering rounding loss

Reply:
My above proposal is not updated `total_stake` and `total_shares` during iteration, in your above example, the nominator will have `2/10`, `3/10`, and `5/10` of the `total_stake` this is also correct and a bit more straightforward IMO, but as I said it is non-blocker.

Reply:
> My above proposal is not updated total_stake and total_shares during iteration, in your above example, the nominator will have 2/10, 3/10, and 5/10 of the total_stake this is also correct and a bit more straightforward IMO

In practically, they will not be same values from the beggining. Will keep it as is for now

Reply:
Rounding loss can be captured by `remaining_stake`.

Reply:
but it is not captured when calculating the nominator share. If that is possible, please give me an example here

Reply:
Yes that is what we get with the current implementation. But you are calculating share_percentage of each nominator after every deposit. You are assuming the share_percentage is not changing which is not true. Here we are calculating nominator's share_percentage after every withdrawal so that (total_shares_left/total_stake_remaining) are upto date instead of pre calculating them earlier to account for those small increments of percentage due to a rounding down in the previous withdrawal

Reply:
> You are assuming the share_percentage is not changing which is not true.

The `share_percentage` is not (and should) changed even in current impl, the only calculation of share is `total_shares = total_shares.checked_sub((&nominator.shares))` which does not bring any rounding loss.

But I do see the difference now, in the current impl nominators can get a cut of the rounding loss. In the above example, after the first nominator is processed, the `total_stake` becomes `80.1` and the second nominator can withdraw `80.1 * 3/8` while my proposal is `100 * 3/10` and leave all rounding loss to treasury.